File: Code/eternal_sim_leader.py
from __future__ import print_function

import argparse
import logging
import os
import shutil
import subprocess
import sys
import time
from random import randint
from functools import partial
from multiprocessing import Manager, Pool


def rngenerator(worker_id, worker_iter, mh, mv, kmax, gamma, measure, n_tunnel_max, lambdascreen, rho_Lambda_thres, fixQ, Nafter, seed, n_recycle, output_dir):
    """
    Generate a random ross n g

    :param int worker_id: The id for this worker
    :param int worker_iter:
    :param float mh:
    :param float mv:
    :param int kmax:
    :param float gamma:
    :param str measure:
    :param int n_tunnel_max:
    :param bool lambdascreen:
    :param float rho_Lambda_thres:
    :param bool fixQ:
    :param float Nafter:
    """
    worker_id = str(worker_id)
    logging.info('worker %s is up and running.' % worker_id)
    worker_outfile = output_dir + ('.worker_%s.txt' % worker_id)

    #call = ['matlab', '-nodisplay', '-nosplash', '-nodesktop', '-nojvm', '-minimize', '-r', '\"test(\'', worker_outfile, '\'),exit\"']
    call = ['sh','/hb/software/apps/matlab/bin/matlab', '-nodisplay','-nosplash','-wait','-nodesktop','-nojvm','-minimize','-r',
        'warning(\'off\');eis_wrapper(\'' +
        worker_outfile + '\',' +
        worker_iter + ',' +
        mv + ',' +
        mh + ',' +
        kmax + ',' +
        gamma + ',' +
        '\'' + measure + '\',' +
        n_tunnel_max + ',' +
        lambdascreen + ',' +
        rho_Lambda_thres + ',' +
        fixQ + ',' +
        Nafter + ',' +
        str(randint(1,1000)) + ',' +
        n_recycle +
        ');exit']
    subprocess.check_call(call)

    logging.info('worker %s received signal to go down.' % worker_id)


def main():
    """
    This will try to run transgene from system arguments
    """
    parser = argparse.ArgumentParser(description=main.__doc__)
    parser.add_argument('--input_file', dest='input_file', type=str, help='Input filename.', required=True)
    parser.add_argument('--cores', dest='cores', type=int, help='Number of cores to use.', required=False, default=1)
    # parser.add_argument('--n_iter', dest='n_iter', type=int, help='Total number of iterations.', required=False, default=10)
    parser.add_argument('--output_file', dest='output_file', type=str, help='Output filename.', required=False, default='outfile.txt')
    parser.add_argument('--output_dir', dest='output_dir', type=str, help='Output directory.', required=False, default='')
    params = parser.parse_args()

    # Default parameter values
    n_iter       = 1000
    mv           = 1.0
    mh           = 1.0
    kmax         = 30
    gamma        = 0.0
    measure      = 'B'
    n_tunnel_max = 3
    lambdascreen = True
    rho_Lambda_thres = 0.0000001
    fixQ         = False
    Nafter       = 60
    seed         = randint(1,1000);
    n_recycle    = 4;

    # Gather parameters from input file
    file_object = open(params.input_file)
    file_contents = file_object.readline().split(",")
    if len(file_contents) >= 1:
        n_iter       = int(  file_contents[0])
    if len(file_contents) >= 2:
        mv           = float(file_contents[1])
    if len(file_contents) >= 3:
        mh           = float(file_contents[2])
    if len(file_contents) >= 4:
        kmax         = int(  file_contents[3])
    if len(file_contents) >= 5:
        gamma        = float(file_contents[4])
    if len(file_contents) >= 6:
        measure      = str(  file_contents[5])
    if len(file_contents) >= 7:
        n_tunnel_max = int(  file_contents[6])
    if len(file_contents) >= 8:
        lambdascreen = int( file_contents[7])
    if len(file_contents) >= 9:
        rho_Lambda_thres = float( file_contents[8])
    if len(file_contents) >= 10:
        fixQ         = int( file_contents[9])
    if len(file_contents) >= 11:
        Nafter       = float(file_contents[10])
    if len(file_contents) >= 12:
        seed         = int(  file_contents[11])
    if len(file_contents) >= 13:
        n_recycle    = int(  file_contents[12])
    print(n_iter,mv,mh,kmax,gamma,measure,n_tunnel_max,
        lambdascreen,rho_Lambda_thres,fixQ,Nafter,seed,n_recycle)

    pool = Pool(processes=params.cores)
    rngenerator_partial = partial(rngenerator,
        worker_iter=str(n_iter/params.cores),
        mh=str(mh),
        mv=str(mv),
        kmax=str(kmax),
        gamma=str(gamma),
        measure=str(measure),
        n_tunnel_max=str(n_tunnel_max),
        lambdascreen=str(lambdascreen),
        rho_Lambda_thres=str(rho_Lambda_thres),
        fixQ=str(fixQ),
        Nafter=str(Nafter),
        seed=str(seed),
        n_recycle=str(n_recycle),
        output_dir=params.output_dir)
    pool.map(rngenerator_partial, range(0, params.cores))
    pool.close()
    pool.join()

    os.system("echo \"Combining output...\"")
    # time.sleep(10)

    # Now that all the processes have completed, we need to process the worker specific vcfs
    worker_files = {i: '.worker_%s.txt' % i for i in range(0, params.cores)}
    with open(params.output_dir + params.output_file, 'w') as outfile:
        for filename in worker_files:
            with open(params.output_dir + worker_files[filename]) as w_file:
                header_line = w_file.readline()
                if filename == 0:
                    print(header_line, file=outfile, end='')
                for line in w_file:
                    print(line, file=outfile, end='')
    for filename in worker_files:
        # Delete the temp files now that we're done with them.
        os.remove(params.output_dir + worker_files[filename])

if __name__ == '__main__':
    main()

File: code/eternal_sim_leader.m

function eternal_sim_leader(cores,flag_range,range,input_file,output_dir)
    
    warning('off');
    
    if nargin < 4, input_file = 'infile.txt';   end
    if nargin < 5, output_dir = './';           end
    
    parpool(cores) % Open a parallel pool
    
    if flag_range
        test_ids = range(1):range(2);
    else
        test_ids = range;
    end
    
    for test_id = test_ids
        
        %% Gather parameters from input file
        
        fid = fopen(sprintf(input_file,test_id),'r');
        meta_line = fgets(fid);
        is = 1; fclose(fid);
        
        [n_iter,~,~,is1] = sscanf(meta_line(is:end),'%E,',1); is = is+is1-1;
        [mv,~,~,is1]     = sscanf(meta_line(is:end),'%G,',1); is = is+is1-1;
        [mh,~,~,is1]     = sscanf(meta_line(is:end),'%G,',1); is = is+is1-1;
        [kmax,~,~,is1]   = sscanf(meta_line(is:end),'%d,',1); is = is+is1-1;
        [gamma,~,~,is1]  = sscanf(meta_line(is:end),'%f,',1); is = is+is1-1;
        
        [measure,~,~,is1]           = sscanf(meta_line(is:end),'%c,',1); is = is+is1-1;
        [n_tunnel_max,~,~,is1]      = sscanf(meta_line(is:end),'%d,',1); is = is+is1-1;
        [lambdascreen,~,~,is1]      = sscanf(meta_line(is:end),'%d,',1); is = is+is1-1;
        [mv_offset_max,~,~,is1]     = sscanf(meta_line(is:end),'%G,',1); is = is+is1-1;
        [aux_flag,~,~,is1]          = sscanf(meta_line(is:end),'%d,',1); is = is+is1-1;
        [Nafter,~,~,is1]            = sscanf(meta_line(is:end),'%G,',1); is = is+is1-1;
        [seed,~,~,is1]              = sscanf(meta_line(is:end),'%d,',1); is = is+is1-1;
        [n_recycle,~,~,is1]         = sscanf(meta_line(is:end),'%d,',1); is = is+is1-1;
        
        if seed == -1
            rs = RandStream.create('mrg32k3a','NumStreams',cores,'Seed','shuffle','CellOutput',true);
        else
            rs = RandStream.create('mrg32k3a','NumStreams',cores,'Seed',seed,'CellOutput',true);
        end
        
        %% Run code in parallel
        
        spmd % Execute this code in parallel
            
            worker_outfile = strcat(sprintf(output_dir,test_id),...
                sprintf('.worker_%d.txt',labindex-1));
            
            warning('off');
            
            eis = EternalInflationSimulator(...
                'outfile',          worker_outfile,...
                'n_iter',           floor(n_iter/cores),...
                'mv',               mv,...
                'mh',               mh,...
                'kmax',             kmax,...
                'gamma',            gamma,...
                'measure',          measure,...
                'n_tunnel_max',     n_tunnel_max,...
                'lambdascreenmode', logical(lambdascreen),...
                'mv_offset_max',    mv_offset_max,...
                'Nafter',           Nafter,...
                'n_recycle',        n_recycle,...
                'randstream',       rs{labindex},...
                'cores',            cores,...
                'flag_screen_basin',aux_flag);
            
            eis.main();
            
        end
        
    end % for

    delete(gcp('nocreate'));
    
end
